plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+	HPI_PORT_REG_EVENT_MASK_BYTE_3,
+	HPI_PORT_REG_SWAP_RESPONSE,
+	HPI_PORT_REG_ACTIVE_EC_MODES,
+	HPI_PORT_REG_VDM_EC_CTRL,
+	HPI_PORT_SPACE_REG_LEN,
+	HPI_PORT_READ_DATA_MEM_ADDR		= 0x400,
+	HPI_PORT_WRITE_DATA_MEM_ADDR		= 0x800,
+} HPIPortReg;
+
+typedef enum {
+	HPI_RESPONSE_NO_RESPONSE,
+	HPI_RESPONSE_SUCCESS			= 0x02,
+	HPI_RESPONSE_FLASH_DATA_AVAILABLE,
+	HPI_RESPONSE_INVALID_COMMAND		= 0x05,
+	HPI_RESPONSE_FLASH_UPDATE_FAILED	= 0x07,
+	HPI_RESPONSE_INVALID_FW,
+	HPI_RESPONSE_INVALID_ARGUMENT,
+	HPI_RESPONSE_NOT_SUPPORTED,
+	HPI_RESPONSE_PD_TRANSACTION_FAILED	= 0x0C,
+	HPI_RESPONSE_PD_COMMAND_FAILED,
+	HPI_RESPONSE_UNDEFINED_ERROR		= 0x0F,
+	HPI_EVENT_RESET_COMPLETE		= 0x80,
+	HPI_EVENT_MSG_OVERFLOW,
+	HPI_EVENT_OC_DETECT,
+	HPI_EVENT_OV_DETECT,
+	HPI_EVENT_CONNECT_DETECT,
+	HPI_EVENT_DISCONNECT_DETECT,
+	HPI_EVENT_NEGOTIATION_COMPLETE,
+	HPI_EVENT_SWAP_COMPLETE,
+	HPI_EVENT_PS_RDY_RECEIVED		= 0x8A,
+	HPI_EVENT_GOTO_MIN_RECEIVED,
+	HPI_EVENT_ACCEPT_RECEIVED,
+	HPI_EVENT_REJECT_RECEIVED,
+	HPI_EVENT_WAIT_RECEIVED,
+	HPI_EVENT_HARD_RESET_RECEIVED,
+	HPI_EVENT_VDM_RECEIVED			= 0x90,
+	HPI_EVENT_SOURCE_CAP_RECEIVED,
+	HPI_EVENT_SINK_CAP_RECEIVED,
+	HPI_EVENT_DP_MODE_ENTERED,
+	HPI_EVENT_DP_STATUS_UPDATE,
+	HPI_EVENT_DP_SID_NOT_FOUND		= 0x96,
+	HPI_EVENT_DP_MANY_SID_FOUND,
+	HPI_EVENT_DP_NO_CABLE_SUPPORT,
+	HPI_EVENT_DP_NO_UFP_SUPPORT,
+	HPI_EVENT_HARD_RESET_SENT,
+	HPI_EVENT_SOFT_RESET_SENT,
+	HPI_EVENT_CABLE_RESET_SENT,
+	HPI_EVENT_SOURCE_DISABLED,
+	HPI_EVENT_SENDER_TIMEOUT,
+	HPI_EVENT_VDM_NO_RESPONSE,
+	HPI_EVENT_UNEXPECTED_VOLTAGE,
+	HPI_EVENT_ERROR_RECOVERY,
+	HPI_EVENT_EMCA_DETECT			= 0xA6,
+	HPI_EVENT_RP_CHANGE_DETECT		= 0xAA,
+	HPI_EVENT_TB_ENTERED			= 0xB0,
+	HPI_EVENT_TB_EXITED
+}HPIResp;
+
+typedef enum
+{
+	/* register region */
+	HPI_REG_PART_REG			= 0,
+	/* data memory for device section */
+	HPI_REG_PART_DATA			= 1,
+	/* flash memory	*/
+	HPI_REG_PART_FLASH			= 2,
+	/* read data memory for port section */
+	HPI_REG_PART_PDDATA_READ		= 4,
+	/* write data memory for port section */
+	HPI_REG_PART_PDDATA_WRITE		= 8
+} HPIRegPart;
+
+typedef struct __attribute__((packed))
+{
+	guint16	event_code;
+	guint16	event_length;
+	guint8	event_data[128];
+}HPIEvent;
+
+static guint16
+fu_ccgx_hpi_reg_addr_gen (guint8 section, guint8 part, guint8 reg_addr)
+{
+	return (guint16) ( (section << 12) | (part << 8) | reg_addr);
+}
+
+static gboolean
+fu_ccgx_hpi_reg_read (FuDevice *device,
+		      CyI2CDeviceHandle *i2c_handle,
+		      guint8 hpi_addr_byte,
+		      guint16 reg_addr,
+		      guint8 *reg_data,
+		      guint16 size,
+		      GError **error)
+{
+	guint32 i;
+	g_autofree guint8 *write_buffer = NULL;
+
+	CyDataBuffer data_buffer = {0};
+	CyI2CDataConfig data_cfg = {0};
+
+	g_return_val_if_fail (i2c_handle != NULL, FALSE);
+	g_return_val_if_fail (reg_data != NULL, FALSE);
+	g_return_val_if_fail (hpi_addr_byte > 0, FALSE);
+
+	write_buffer = g_malloc0 (hpi_addr_byte + 1);
+
+	g_return_val_if_fail (write_buffer != NULL, FALSE);
```

Comment:
I think this specific `g_return_val_if_fail` is probably overkill.

Reply:
Is it right for me to remove g_return_val_if_fail and not to check whether write_buffer is NULL?

Reply:
I think g_return_val_if_fail() can be removed quite safely.